### DISCUSSION TITLE
How to assemble these images?

### DISCUSSION THREAD

**Discussion by DJX995** [ImageMagick/ImageMagick#5506] [Help] [answered]
I have a group of images that are assembled from tiles of varying resolutions in a grid to a final large image. I actually have 65 of these groups. I had been assembling them by hand in mspaint but I can't complete them because some of the resolutions are too large.

The images are named like this:
n-nn-nn.jpg
First n is a single digit magnification level
Second n is either a single or double digit number and is the tile X position
Third n is either a single or double digit number and is the tile Y position

For example, this is how you would assemble an image
5-0-0.jpg:5-1-0.jpg:etc:etc:5-15-0.jpg5-16-0.jpg
5-0-1.jpg:5-1-1.jpg
5-0-2.jpg:5-1-2.jpg
etc
etc
5-0-16.jpg:5-1-16.jpg

The tiles are not all the same size as when you get close to the edges of the image, the tiles are whatever is left over to complete the image in that row or column. The image groups are all described in an XML file with the following information:
```
<?xml version="1.0"?>
<IMAGE_PROPERTIES TILESIZE="256" VERSION="1.8" NUMIMAGES="1" NUMTILES="312" HEIGHT="2850" WIDTH="4800"/>
```

Any help is appreciated.
Thanks!

**Comment by snibgo** [comment]
What version of IM, on what platform (what shell)?

On each image, do you want gaps between tiles? Does each row of tiles in an image have the same number of tiles? Are they in neat rows and columns?

How large is each assembled image (in pixels)? Do you have enough memory to fit all the required tiles, and the output image, at the same time?

Is any metadata from your XML file needed to assemble the tiles?

Possible methods include:

"`magick montage`" might work well if each image has a simple assembly.

You might create a blank canvas, and composite each tile with the appropriate geometry.

You might append tiles into a single row, then append another set of tiles into a second row, and so on, finally appending all the rows together.

**Comment by DJX995** [comment]
1. ImageMagick-7.1.0-47-Q16-x64-static.exe/Win64
2. No gaps, they are in neat rows and columns. If they are aligned properly, they should be exactly adjacent to one another.
3. The largest resolution is specified in the XML for each image. I was using this as a starting point in mspaint by creating a canvas with that resolution and then copying and pasting the tiles one at a time. The resolutions got too great that the time required and precision of my hand was no longer feasible. I don't think I would run out of memory with 32GB.
4. The XML contains the final resolution for the image.

I'm new to this program but I've used ffmpeg quite a bit. Not sure if this is even possible.
Thanks.

**Comment by snibgo** [comment]
I would do this by building an IM command that looked something like this:
```
magick ^
  ( 5-0-0.jpg 5-1-0.jpg ... 5-15-0.jpg 5-16-0.jpg ^
    +append +repage ) ^
  ( 5-0-1.jpg 5-1-1.jpg ... 5-15-1.jpg 5-16-1.jpg ^
    +append +repage ) ^
  :
  :
  ( 5-0-16.jpg 5-1-16.jpg ... 5-15-16.jpg 5-16-16.jpg ^
    +append +repage ) ^
  -append +repage ^
  out.tiff
```
Does that command do what you want?

If it does, then the next step is to automatically build that command. This would be a BAT script that writes another script that contains the command, and executes that other script. 

The main script could loop through all possible values for y. Then a nested loop through all possible values for x. It would echo filenames and other elements into the other BAT file.

If all your images have 17x17 tiles, the job is simple. Otherwise, the main script would need to figure out how many tile are in each row, and how many rows in each image. It could figure this out from whether files exist.

**Comment by DJX995** [comment]
Yes, your command produces what I need.
`"C:\Program Files\ImageMagick-7.1.0-Q16\magick.exe" ( 5-0-0.jpg 5-1-0.jpg 5-2-0.jpg 5-3-0.jpg 5-4-0.jpg 5-5-0.jpg 5-6-0.jpg 5-7-0.jpg 5-8-0.jpg 5-9-0.jpg 5-10-0.jpg 5-11-0.jpg 5-12-0.jpg 5-13-0.jpg 5-14-0.jpg 5-15-0.jpg 5-16-0.jpg 5-17-0.jpg 5-18-0.jpg +append +repage ) ( 5-0-1.jpg 5-1-1.jpg 5-2-1.jpg 5-3-1.jpg 5-4-1.jpg 5-5-1.jpg 5-6-1.jpg 5-7-1.jpg 5-8-1.jpg 5-9-1.jpg 5-10-1.jpg 5-11-1.jpg 5-12-1.jpg 5-13-1.jpg 5-14-1.jpg 5-15-1.jpg 5-16-1.jpg 5-17-1.jpg 5-18-1.jpg +append +repage ) -append +repage out.tiff`
This produces the first two rows completely assembled.
Only 11 more rows to go, haha.

All images are different, they have a varying number of tiles (resolutions).

**Reply by fmw42** [reply]
Can't you just use montage?

**Reply by DJX995** [reply]
> Can't you just use montage?

I couldn't figure out the syntax. This is a new program for me.

**Comment by fmw42** [comment]
`magick montage image1.suffix image2.suffix ... imageN.suffix (or *.suffix) -tile Number_ColsxNumber_Rows -geometry +0+0 output.suffix`

**Reply by DJX995** [reply]
This produces an image that is randomly assembled with extra white space on the end tiles (wrong resolution).

**Reply by fmw42** [reply]
It is only "random" in your view because you have not sorted your names appropriately. Montage takes all files (*.suffix) in alphabetic ordering, not numerical. It does not know about row column indices.  The extra space will be apparent if your images are not the same dimensions. But in that case, it will be the same if using +-append. You can crop off the extra space as post processing in either case.  If you are happy with +-append, then go with it. If you order your images manually in montage the same way you order them in append, then the result should be similar -- not "random".

**Comment by snibgo** [comment]
The tricky part to this task isn't the IM command, but getting the filenames in the correct order. When that is done, the IM command could use append, or montage, or whatever.

Montage will use files in alphabetical order. If the names were formatted like `5-YY-XX.jpg` where YY and XX had leading zeros, the task would be much simpler.

**Reply by DJX995** [reply]
I understand. I'm just working with what I got. Some program created these, I'm just trying to put them back together.  I appreciate your help. Assembling the commands will take some time by hand but not nearly as long as assembling the images by hand like I started doing. Thank you.

**Reply by fmw42** [reply]
You could script to rename your files so that any single digit was made into a leading 0 and the the digit. Or find a renaming application. There are many of them or perhaps one that is online.

**Comment by snibgo** [comment, marked as answer]
> Assembling the commands will take some time by hand...

Why do them by hand? A couple of `for` loops will do the job, like this Windows BAT script:
```
echo off
del x.txt
(
  echo magick ^^
  for /L %%Y in (0,1,16) do (
    echo ^( ^^
    for /L %%X in (0,1,16) do (
      echo 5-%%X-%%Y.jpg ^^
    )
    echo +append +repage ^) ^^
  )
  echo -append +repage out.tiff
) >x.bat
```
The file `x.bat` contains a single magick command. Adjust each "16" as required. Or make them environment variables.

**Reply by DJX995** [reply]
Thank you!
`echo off
del x.bat
(
  echo "C:\Program Files\ImageMagick-7.1.0-Q16\magick.exe" ^^
  for /L %%Y in (0,1,11) do (
    echo ^( ^^
    for /L %%X in (0,1,18) do (
      echo 5-%%X-%%Y.jpg ^^
    )
    echo +append +repage ^) ^^
  )
  echo -append +repage out.tiff
) >x.bat`

`x.bat`

Produces the entire image for tile set 02!

Thanks again!

**Comment by snibgo** [comment]
My script has a bug. Near the top, change `del x.txt` to `del x.bat`.

**Reply by DJX995** [reply]
Very minor, I understand what you're doing. Thanks again!

**Comment by DJX995** [comment]
My only other question is why did you choose TIFF as the output format over JPG (the files are originally JPG)?
I would like to preserve original quality if at all possible.

**Comment by snibgo** [comment]
You can use any format you want, of course. When an image is written to JPG, there is lossy compression, so some pixel values will change, which will lose some quality. This may not matter.
